### PR TITLE
Narrow virt-handler permissions

### PR DESCRIPTION
**Pull request by mlsorensen**
**What this PR does / why we need it**:

Allows us to remove cluster admin privileges from virt-handler.

**Special notes for your reviewer**:
I'm not sure if this is the right place to edit the RBAC, or if these are in fact generated from something else as the directory implies.

I tested this RBAC with 'make functest' and got the same results pre and post change.

**Release note**:

```release-note
add 'virt-handler' service account and limit privileges for virt-handler
```

### CODE REVIEW

File: manifests/generated/rbac.authorization.k8s.yaml.in [hunk 1]
```diff
@@ -406,3 +406,96 @@ subjects:
...
+  resources:
+  - secrets
+  - persistentvolumeclaims
+  verbs:
+  - get
+- apiGroups:
+  - ""
+  resources:
+  - nodes
+  - events
```
Comment (cynepco3hahue) [inline manifests/generated/rbac.authorization.k8s.yaml.in]: duplication of events resources
line 444 and 450
I think you can just remove line 444

File: manifests/generated/rbac.authorization.k8s.yaml.in [hunk 2]
```diff
@@ -406,3 +406,96 @@ subjects:
...
+    kubevirt.io: ""
+  name: kubevirt-handler
+  namespace: {{.Namespace}}
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  labels:
+    kubevirt.io: ""
+  name: kubevirt-handler
```
Comment (cynepco3hahue) [inline manifests/generated/rbac.authorization.k8s.yaml.in]: can you rename all `kubevirt-handler` to `virt-handler`?

File: manifests/generated/virt-handler.yaml.in [hunk 3]
```diff
@@ -60,7 +60,7 @@ spec:
         - mountPath: /var/lib/kubelet/device-plugins
           name: device-plugin
       hostPID: true
-      serviceAccountName: kubevirt-privileged
+      serviceAccountName: kubevirt-handler
```
Comment (cynepco3hahue) [inline manifests/generated/virt-handler.yaml.in]: same here just rename to `virt-handler`

Comment (mlsorensen) [reply manifests/generated/virt-handler.yaml.in]: I was attempting to be consistent with the kubevirt-apiserver and kubevirt-controller service accounts. I'd prefer not to unnecessarily change those two in this pull to maintain consistency with a 'virt-handler' rename, but I'll do what the community wants.

File: manifests/generated/rbac.authorization.k8s.yaml.in (hunk 1 above)
Comment (mlsorensen) [reply manifests/generated/rbac.authorization.k8s.yaml.in]: will fix

**Comment by mlsorensen** [comment]
> We generate our manifest via code, so you will need to change the code under `pkg/virt-operator/creation/` and run `make generate` to get updated manifests.

Ok, I'll look at this. I thought pkg/virt-operator/creation was specific to the operator, and I didn't see any handler permissions in there. I'll look again.

> 
> Also, I do not sure that we do not need privilege rights for the virt-handler, because it can be some corner cases like SRIOV usage when it still will need privilege rights.

Certainly it doesn't need kubernetes cluster admin to do things related to SR-IOV, right?  It may need to be a privileged container but that doesn't have any relation to RBAC.

**Comment by mlsorensen** [comment]
I've added the manifest generation, and it generated the same as what I had manually put into "manifests/generated/rbac.authorization.k8s.yaml.in".

Please review and let me know if we should keep 'kubevirt-handler' service account naming, or rename to 'virt-handler' along with renaming 'kubevirt-apiserver' to 'virt-apiserver', 'kubevirt-controller' to 'virt-controller' and 'kubevirt-privileged' to 'virt-privileged' for consistency.

**Comment by slintes** [comment]
> Maybe something specific to OpenShift, because we also add that account to the privileged SCC?

looking at the tests, that might be indeed the case, openshift lanes have many failures...
http://jenkins.ovirt.org/job/kubevirt_kubevirt_standard-check-pr/3300/artifact/ci_build_summary.html

will trigger another CI run

**Comment by slintes** [comment]
ci test please

**Comment by mlsorensen** [comment]
> As far as I can see, `virt-handler` was the only user of the `kubevirt-privileged` account. So I guess it's safe to delete it.

I thought the operator used the kubevirt-privileged account. I don't use the operator (yet) so it isn't an immediate concern for me, but I assume that will eventually need to be gone through as well.

**Comment by mlsorensen** [comment]
> @mlsorensen I believe we need to add correct `scc` to the new service account, please check `deploy.sh` script.
> 
> ```
> if [[ "$KUBEVIRT_PROVIDER" =~ os-* ]]; then
>     _kubectl create -f ${MANIFESTS_OUT_DIR}/testing/ocp
>     _kubectl adm policy add-scc-to-user privileged -z kubevirt-controller -n ${namespace}
>     _kubectl adm policy add-scc-to-user privileged -z kubevirt-privileged -n ${namespace}
>     _kubectl adm policy add-scc-to-user privileged -z kubevirt-apiserver -n ${namespace}
>     # Helpful for development. Allows admin to access everything KubeVirt creates in the web console
>     _kubectl adm policy add-scc-to-user privileged admin
> fi
> ```
> We also need to be aware that this change will affect kubevirt-ansible repository.

it looks like that script was overhauled Monday to only have kubevirt-operator.

**Comment by mlsorensen** [comment]
> > As far as I can see, `virt-handler` was the only user of the `kubevirt-privileged` account. So I guess it's safe to delete it.
> 
> I thought the operator used the kubevirt-privileged account. I don't use the operator (yet) so it isn't an immediate concern for me, but I assume that will eventually need to be gone through as well.

I don't know where I saw that, as I can't find it now.  I guess I'll remove the privileged account as a part of this change.

**Comment by slintes** [comment]
> I thought the operator used the kubevirt-privileged account.

The operator has it's own `kubevirt-operator` account, which is bound to the cluster-admin role. That one isn't  generated. See `manifests/release/kubevirt-operator.yaml.in`.
I never thought about generating that manifest too, because the operator doesn't need it itself. We moved all other manifests into code, because the operator will need the code, and we didn't want to maintain it at 2 places.

**Comment by slintes** [comment]
> it looks like that script was overhauled Monday to only have kubevirt-operator

ah yes, you're right. We removed the old deployment method...

**Comment by mlsorensen** [comment]
Regarding why it was cluster admin, I assume it was just an artifact of the initial development, as kubevirt-controller was admin as well but recently refactored in #1166. If there's more to this please let me know. The handler RBAC was just brought up to me as a security issue that I'm trying to resolve.

File: manifests/generated/virt-handler.yaml.in (hunk 3 above)
Comment (davidvossel) [reply manifests/generated/virt-handler.yaml.in]: yeah, the way you have it now is consistent with our other service accounts. for the sake of this PR I think we can continue with kubevirt-handler as the SA's name.

File: pkg/virt-operator/util/client.go [hunk 4]
```diff
@@ -130,7 +130,7 @@ func UpdateScc(clientset kubecli.KubevirtClient, kv *virtv1.KubeVirt, add bool)
 
 	var kubeVirtAccounts []string
 	prefix := "system:serviceaccount"
-	kubeVirtAccounts = append(kubeVirtAccounts, fmt.Sprintf("%s:%s:%s", prefix, kv.Namespace, "kubevirt-privileged"))
+	kubeVirtAccounts = append(kubeVirtAccounts, fmt.Sprintf("%s:%s:%s", prefix, kv.Namespace, "kubevirt-handler"))
```
Comment (davidvossel) [inline pkg/virt-operator/util/client.go]: Oh i missed this. You already take care of setting the scc. Ignore my previous comment then

**Comment by davidvossel** [comment]
> Also, I do not sure that we do not need privilege rights for the virt-handler, because it can be some corner cases like SRIOV usage when it still will need privilege rights.

@booxter do you know if this change will break SRIOV?

**Comment by cynepco3hahue** [comment]
ci test please

**Comment by cynepco3hahue** [comment]
I see some strange errors under one of jobs https://jenkins.ovirt.org/job/kubevirt_kubevirt_standard-check-pr/3351//artifact/check-patch.k8s-1.10.11.el7.x86_64/mock_logs/script/stdout_stderr.log
```
Unexpected Warning event received: testvmiklzrftmdwz7npbzlwzz4nwsvcbqr5j48hv4qbg4dwx8rcj9s,bd7931e5-23fc-11e9-8c1f-525500d15501: server error. command Launcher.Sync failed: "LibvirtError(Code=1, Domain=10, Message='internal error: qemu unexpectedly closed the monitor: 2019-01-29T19:33:37.502746Z qemu-system-x86_64: -drive file=/var/run/kubevirt-ephemeral-disks/container-disk-data/kubevirt-test-default/testvmiklzrftmdwz7npbzlwzz4nwsvcbqr5j48hv4qbg4dwx8rcj9s/disk_disk0/disk-image.qcow2,format=qcow2,if=none,id=drive-ua-disk0,cache=none: Image is not in qcow2 format')"
```
but it fails only under one of lanes, so possibly it does not relate to the PR.

I want to run CI again, and check if we will have same errors under the same lane.

**Comment by cynepco3hahue** [comment]
ci test please

**Comment by cynepco3hahue** [comment]
@mlsorensen Can you please rebase?

**Comment by davidvossel** [comment]
#1970 SRIOV is making it's way into CI soon, so once this makes it in we'll be able to validate this PR doesn't impact functionality.

**Comment by cynepco3hahue** [comment]
ci test please

**Comment by cynepco3hahue** [comment]
ci test please

**Comment by cynepco3hahue** [comment]
@davidvossel @slintes I think it looks good to go, does someone have any objections?

**Comment by davidvossel** [comment]
ci test please

**Comment by davidvossel** [comment]
I think this is good to go. I just want to run CI one more time
